server/chat-plugins/repeats.ts

```diff
@@ -0,0 +1,166 @@
+/**
+ * Chat plugin for repeating messages in chat
+ * Based on bot functionality from Kid A and Expecto Botronum
+ * @author Annika
+ */
+
+export interface RepeatedPhrase {
+	phrase: string;
+	/** interval in milliseconds */
+	interval: number;
+}
+
+/** roomid:phrase:timeout map */
+export const repeats = new Map<RoomID, Map<string, NodeJS.Timeout | null>>();
+
+export function hasRepeat(room: BasicRoom, phrase: string) {
+	return !!repeats.get(room.roomid)?.get(phrase);
+}
+
+export function addRepeat(room: BasicRoom, repeat: RepeatedPhrase) {
+	if (!room.settings.repeats) room.settings.repeats = [];
+	room.settings.repeats.push(repeat);
+	room.saveSettings();
+	runRepeat(room, repeat);
+}
+
+export function removeRepeat(room: BasicRoom, phrase: string) {
+	if (!room.settings.repeats) return;
+	room.settings.repeats = room.settings.repeats.filter(repeat => repeat.phrase !== phrase);
+	room.saveSettings();
+
+	if (!repeats.has(room.roomid)) return;
+	const oldInterval = repeats.get(room.roomid)?.get(phrase);
+	if (oldInterval) clearInterval(oldInterval);
+	repeats.get(room.roomid)?.set(phrase, null);
+}
+
+function runRepeat(room: BasicRoom, repeat: RepeatedPhrase) {
+	if (!repeats.has(room.roomid)) {
+		repeats.set(room.roomid, new Map<string, NodeJS.Timeout | null>());
+	}
+
+	const displayRepeat = (targetRoom: BasicRoom, phrase: string) => {
+		targetRoom.add(`|html|<div class="infobox">${phrase}</div>`);
+		targetRoom.update();
+	};
```

Comment:
I don't like binding parameters because it does the same thing as having them in closures, but worse.

I would actually make `Repeats` into an entire class, if I were you. Do you mind if I refactor this?

Reply:
But if I were to refactor only this function, it would look like:

```ts
function runRepeat(room: BasicRoom, repeat: RepeatedPhrase) {
    const roomid = room.roomid;
    let roomRepeats = repeats.get(roomid);
    if (!roomRepeats) {
        roomRepeats = new Map();
        roomRepeats.set(roomid, roomRepeats);
    }

    if (roomRepeats.has(repeat.phrase)) {
        handle this case somehow
    }

	roomRepeats.set(
		repeat.phrase,
		setInterval(() => {
            if (room !== Rooms.get(room.roomid)) {
                handle room destroyed
                return;
            }
            room.add(`|html|<div class="infobox">${repeat.phrase}</div>`);
            room.update();
        }, repeat.interval)
	);
}
```

Reply:
If you want to refactor this you can; if not I can remove the bound parameters thing and fix the memory leak.

Reply:
@Zarel bump?